implement rate limited queue to control Certificate creation rate

hanapedia opened the pull request:
## Problem addressed in this PR
The current implementation of contour-plus provides no mechanism to control the rate at which Certificate resources are created. This can be problematic when a large number of Certificates need to be issued or re-issued in a short period of time, as it may trigger rate limits enforced by public CAs.

cert-manager itself also does not provide a way to directly control the rate at which Certificate issuance requests are sent to upstream CAs. As a result, the only practical control point available to contour-plus is the rate at which it creates Certificate resources.

## Solution
This PR introduces a simple queue + worker model in contour-plus to control the rate of Certificate resource creation.
A configurable rate limiter is applied to the worker responsible for creating Certificates, allowing operators to throttle Certificate issuance requests before they reach cert-manager and the upstream CA.

It is implemented using the rate limiter and workqueue provided in client-go packages.

The rate limit is configured via the command-line flag: `--certificate-apply-limit`
- The value is a float64 representing the maximum number of Certificate creations allowed per second.
- Passing a negative value results in an error.
- The feature is disabled by default (--certificate-apply-limit=0), in which case contour-plus behaves as before:
  - Certificates are created immediately
  - No queue or rate limiting is applied

## Some of the design choices I made
- Unified applier interface
The logic for applying Certificate resources is abstracted behind a common interface, allowing both queued (rate-limited) and non-queued (immediate) implementations to share the same call path.
- Controller-managed lifecycle
The workqueue–based applier is started during controller setup and registered as a runnable with the manager, ensuring that its lifecycle is tied to the controller and properly handled on startup and shutdown.
~~- Test visibility into apply path~~
~~The test code wraps the applier implementation to expose whether a Certificate was applied via the queue or directly, allowing tests to assert the correct behavior without coupling to internal implementation details.~~
  - removed the integration test with wrapper and refactored the majority of the new tests as unit test.

hanapedia committed:
implement rate limited workqueue for Certificate apply

Signed-off-by: Hiroki Hanada <hiroki-hanada@cybozu.co.jp>

hanapedia committed:
add tests for cert apply worker

Signed-off-by: Hiroki Hanada <hiroki-hanada@cybozu.co.jp>

hanapedia committed:
fix annotation/label propagation requeue check

Signed-off-by: Hiroki Hanada <hiroki-hanada@cybozu.co.jp>

chez-shanpu commented on controllers/certificate_worker.go:
```diff
@@ -0,0 +1,182 @@
+				return
+			}
+
+			if err := applyCertificate(ctx, w.client, obj); err != nil {
```
I think we should check the context before `applyCertificate` here because it can be reached even after `ctx.Done()` has been closed.
We can check if the `ctx.Done` is closed like the following

```go
  if ctx.Err() != nil {
      log.Info("context cancelled, skipping cert apply", "key", objKey.String())
      return
  }
```

chez-shanpu commented on controllers/certificate_worker.go:
```diff
@@ -0,0 +1,182 @@
+
+			w.mu.Lock()
+			defer w.mu.Unlock()
+			obj, ok := w.manifests[objKey]
```
nit: It seems this lock covers a broader range than necessary, which could be improved for better performance. Currently, `applyCertificate` is included within the lock, but that shouldn't be necessary.

I think we can get the lock only for the necessary sections, like this
```go
			w.mu.Lock()
			obj, ok := w.manifests[objKey]
			w.mu.Unlock()
			if !ok {...}
```

chez-shanpu reviewed (commented):
This is just an initial review of the non-test files :)

hanapedia commented on controllers/certificate_worker.go:
```diff
@@ -0,0 +1,182 @@
+			if err := applyCertificate(ctx, w.client, obj); err != nil {
+				log.Error(err, "cert apply failed", "key", objKey.String())
+				w.workqueue.AddRateLimited(objKey)
+				return
```
Also, we are using BucketRateLimiter for a global rate limit, and it seems that we cannot use `.NumRequeues` unlike other per-key rate limiters. It always returns 0 for BucketRateLimiter.
https://github.com/kubernetes/client-go/blob/d8ad3ab106f8db053c719f8973637da88664ebb9/util/workqueue/default_rate_limiters.go#L72-L74

chez-shanpu replied:
> we could potentially remove the retry logic entirely

> I'd suggest deferring that to another issue rather than addressing it in this PR.

That makes sense. In the non-queued case, when an error occurs, it gets re-added to the Reconcile queue, so retries happen automatically. Given the `.NumRequeues` behavior you mentioned (always returning 0 for BucketRateLimiter), I agree, it might be better to print an error log and return without retrying.

---
This is just one Idea, but if we implement retries when using the queue, we could have the enqueuing goroutine wait for the result via a channel, and have `Start` send the error using that channel. This way, we could use the Reconcile queue for retries like the non-queued case.

chez-shanpu replied:
I think blocking the reconciliation is technically fine because `Reconcile` run serially for the same resource but can execute in parallel for different resources. But that said, your proposed approach seems better :)

hanapedia committed:
reflect review comments

Signed-off-by: Hiroki Hanada <hiroki-hanada@cybozu.co.jp>

hanapedia committed:
implement retry for the apply worker

Signed-off-by: Hiroki Hanada <hiroki-hanada@cybozu.co.jp>

chez-shanpu commented on controllers/certificate_worker.go:
```diff
@@ -0,0 +1,235 @@
+			Name: "certificate-apply",
+		},
+	)
+	retryCh := make(chan event.TypedGenericEvent[*projectcontourv1.HTTPProxy])
```
It would be better to add a buffer to `retryCh` to avoid blocking in `enqueueHTTPProxy`.

```suggestion
	retryCh := make(chan event.TypedGenericEvent[*projectcontourv1.HTTPProxy], 10)
```

chez-shanpu commented on controllers/certificate_worker.go:
```diff
@@ -0,0 +1,235 @@
+	}); err != nil {
+		return err
+	}
+	return nil
```
```suggestion
	return k8sClient.Patch(ctx, obj, client.Apply, &client.PatchOptions{
		Force:        ptr.To(true),
		FieldManager: "contour-plus",
	})
```
nit

chez-shanpu reviewed (commented):
The retry implementation looks good :D
I left a few small comments.

hanapedia committed:
reflect review comments

Signed-off-by: Hiroki Hanada <hiroki-hanada@cybozu.co.jp>

chez-shanpu reviewed (dismissed):
LGTM

hanapedia committed:
add limiter value validation in ctor

Signed-off-by: Hiroki Hanada <hiroki-hanada@cybozu.co.jp>

hanapedia committed:
rework cert worker test

Testing queuing behaviors of certificate apply worker integrated in
the main reconcile loop was prone to flakiness.
This commit reworked the test for the queueing behavior of cert apply worker
to be independent of the main reconcile loop.
Also, this commit moved the integration test for the happy path to httpproxy_controller_test.

Signed-off-by: Hiroki Hanada <hiroki-hanada@cybozu.co.jp>

chez-shanpu commented on controllers/certificate_worker_test.go:
```diff
@@ -0,0 +1,331 @@
+		ctx    context.Context
+	)
+
+	BeforeEach(func() {
```
nit We may be able to use `BeforeAll`.

chez-shanpu commented on controllers/certificate_worker_test.go:
```diff
@@ -0,0 +1,331 @@
+		// This allows the inspection of the queue content withouot adding a wrapper.
+		var key types.NamespacedName
+
+		BeforeEach(func() {
```
diito

chez-shanpu commented on controllers/certificate_worker_test.go:
```diff
@@ -0,0 +1,331 @@
+				},
+			}
+
+			// Call Apply, which internally calls RequiresQueue and should enqueue
```
```suggestion
			By("calling apply, which internally calls RequiresQueue and should enqueue")
```
info You can use `By()` to add explanations for things that don't need a block.

chez-shanpu reviewed (commented):
I left some small comments.

hanapedia committed:
reflect review comments

Signed-off-by: Hiroki Hanada <hiroki-hanada@cybozu.co.jp>

chez-shanpu commented on controllers/setup.go:
```diff
@@ -28,29 +29,47 @@ type ReconcilerOptions struct {
+}
+
+// SetupAndGetReconciler initializes reconcilers and return the reconciler struct
+// c
```
```suggestion

```

chez-shanpu reviewed (commented):
I think this PR is almost ready to merge.
I left a few minor comments.

hanapedia committed:
reflect more review comments

Signed-off-by: Hiroki Hanada <hiroki-hanada@cybozu.co.jp>

chez-shanpu reviewed (dismissed):
LGTM

hanapedia committed:
add metrics for counting queued applies (#132)

Signed-off-by: Hiroki Hanada <hiroki-hanada@cybozu.co.jp>
